### PR TITLE
feat: Add ut to all helper functions

### LINKED ISSUE #54: Add unit tests for all helper functions
All helper functions or those functions that do calculations etc should have unit tests. I.e. where it makes sense.

### PR DESCRIPTION
**Pull request by apollo-pablo**
Closes #54 

> Problem

We dont have any test on math functions.

> Tasks

- [x] Add UT to implementations Helpers functions.
- [x] Add on every file instead of new files.

> Notes

Format and lints failures are from develop branch, we should make a PR and fix them first.

### CODE REVIEW

File: src/implementations/astroport/helpers.rs [hunk 1]
```diff
@@ -132,11 +149,24 @@ pub(crate) fn compute_d(leverage: u64, amount_a: u128, amount_b: u128) -> Option
...
+                Some(new_d) => {
+                    // Equality with the precision of 1
+                    if new_d == d {
+                        break;
+                    }else{
+                        new_d
+                    }
+                },
+                None => break,
+            };
```
Comment (apollo-pacman) [inline src/implementations/astroport/helpers.rs:163]: why did we need to make this change?

**Review by apollo-pacman — COMMENTED** [review]
nice work! 👍 a few comments @apollo-pablo

File: src/implementations/junoswap/helpers.rs [hunk 2]
```diff
@@ -295,3 +305,138 @@ pub(crate) fn juno_simulate_provide_liquidity(
...
+    //                     next_amp,
+    //                     next_amp_time,
+    //                 };
+    
+    //         // When
+    //         compute_current_amp(&config, block_time)?;
+    
+    //         // Then Should not panic
+    //     }
+    // }
```
Comment (apollo-sturdy) [inline src/implementations/junoswap/helpers.rs:441]: Remove unused code.

File: src/implementations/astroport/helpers.rs [hunk 3]
```diff
@@ -133,10 +132,10 @@ pub(crate) fn compute_d(leverage: u64, amount_a: u128, amount_b: u128) -> Option
             //d = (leverage * sum_x + d_p * n_coins) * d / ((leverage - 1) * d + (n_coins +
             // 1) * d_p);
             d = calculate_step(&d, leverage, sum_x, &d_product)?;
-            // Equality with the precision of 1
+                    // Equality with the precision of 1
             if d == d_previous {
-                break;
-            }
+                        break;
+                    }
```
Comment (apollo-sturdy) [inline src/implementations/astroport/helpers.rs:138]: Why add extra spacing? Pls use rustfmt.

File: src/implementations/junoswap/helpers.rs [hunk 4]
```diff
@@ -227,6 +229,13 @@ pub(crate) struct JunoProvideLiquidityInfo {
 /// ### Returns
 /// The amount of token1 and token2 that should be sent to Junoswap to provide
 /// liquidity and the expected amount of lp tokens to be minted.
+/// 
+/// ### Requirements
+/// - pool_info.token2_reserve > 0
+/// - token2.amount > 0
+/// 
+/// ### Note
+/// This function does not take into account protocol_fee_percent and lp_fee_percent
```
Comment (apollo-sturdy) [inline src/implementations/junoswap/helpers.rs:238]: Neither does Junoswap, so why mention it? Please remove comment.

File: src/implementations/junoswap/helpers.rs [hunk 5]
```diff
@@ -227,6 +229,13 @@ pub(crate) struct JunoProvideLiquidityInfo {
 /// ### Returns
 /// The amount of token1 and token2 that should be sent to Junoswap to provide
 /// liquidity and the expected amount of lp tokens to be minted.
+/// 
+/// ### Requirements
+/// - pool_info.token2_reserve > 0
+/// - token2.amount > 0
```
Comment (apollo-sturdy) [inline src/implementations/junoswap/helpers.rs:235]: Not sure what this means. Can you please either remove this or explain it further so its understandable?

File: src/implementations/junoswap/helpers.rs [hunk 6]
```diff
@@ -295,3 +305,138 @@ pub(crate) fn juno_simulate_provide_liquidity(
...
+            info: JunoAssetInfo(Denom::Cw20(Addr::unchecked("dai"))),
+            amount: Uint128::from(amount_b),
+        };
+        let assets: JunoAssetList = JunoAssetList(vec![usdc.to_owned(), dai.to_owned()]);
+        let pool_info: InfoResponse = InfoResponse {
+            token1_reserve: Uint128::from(reserve_a),
+            token1_denom: usdc.info.0,
+            token2_reserve: Uint128::from(reserve_b),
+            token2_denom: dai.info.0,
+            lp_token_supply: usdc.amount + dai.amount, // should this be reserve_a+reserve_b?
```
Comment (apollo-sturdy) [inline src/implementations/junoswap/helpers.rs:345]: Shouldn't matter what this value is. Remove comment.

File: src/implementations/junoswap/helpers.rs [hunk 7]
```diff
@@ -295,3 +305,138 @@ pub(crate) fn juno_simulate_provide_liquidity(
...
+    
+    
+    // Edge borders testing
+    #[test_case(100,0,100,1 => matches Err(_); "when reserve_a is zero should err")]
+    #[test_case(100,1,100,0 => matches Err(_); "when reserve_b is zero should err FAILS")]
+    #[test_case(10000000,1,0,1 => matches Err(_); "when amount_b is zero should err FAILS")]
+    #[test_case(0,1,0,1 => Ok( (0,0,0) ); "when amount_a and amount_b zero should work")]
+    #[test_case(30,1,1,1 => with |i: Result<(u128,u128,u128),StdError> | assert!(i.unwrap().2 == 1u128); "when asset_ratio gt pool_ratio with amount_a gt amount_b")] // TODO: what happens with the rest of amount_a? 
+    #[test_case(1,2,1,2 => Ok( (1,2,1) ); "when pool_ratio greater than asset_ratio")]
+    fn juno_simulate_provide_liquidity_test(
```
Comment (apollo-sturdy) [inline src/implementations/junoswap/helpers.rs:325]: Some of these tests fail. Please fix before request review.

**Review by apollo-sturdy — CHANGES_REQUESTED** [review]

**Pull request closed without merging**